rs/xnet/endpoint/src/lib.rs

```diff
@@ -170,114 +293,24 @@ impl XNetEndpoint {
-
-                            async move {
-                                let _ = &ctx;
-                                let (response_sender, response_receiver) = oneshot::channel();
-                                let task = WorkerMessage::HandleRequest {
-                                    request,
-                                    response_sender,
-                                };
-
-                                // NOTE: we must use non-blocking send here, otherwise we might
-                                // delay the event thread.
-                                if ctx.request_sender.try_send(task).is_err() {
-                                    ctx.metrics
-                                        .request_duration
-                                        .with_label_values(&[
-                                            RESOURCE_UNKNOWN,
-                                            StatusCode::SERVICE_UNAVAILABLE.as_str(),
-                                        ])
-                                        .observe(0.0);
-
-                                    return ok(Response::builder()
-                                        .status(StatusCode::SERVICE_UNAVAILABLE)
-                                        .body(Body::from("Queue full"))
-                                        .unwrap());
-                                }
-
-                                ok(response_receiver.await.unwrap_or_else(|e| {
-                                    panic!("XNet Endpoint Handler shut down unexpectedly: {}", e)
-                                }))
-                            }
-                        }
-                    }))
-                }
-            }
-        });
-
-        let (address, server) = {
-            let _guard = runtime_handle.enter();
-
-            #[cfg(test)]
-            use ic_xnet_hyper::tls_bind_for_test as tls_bind;
-
-            #[cfg(not(test))]
-            use ic_xnet_hyper::tls_bind;
-
-            let (addr, builder) =
-                tls_bind(&config.address, tls, registry_client).unwrap_or_else(|e| {
-                    panic!(
-                        "failed to bind XNet socket, address {:?}: {}",
-                        config.address, e
-                    )
-                });
-            (
-                addr,
-                builder
-                    .executor(ExecuteOnRuntime(runtime_handle.clone()))
-                    .serve(make_service),
-            )
+        let ctx = Context {
+            metrics: Arc::clone(&metrics),
+            request_sender: request_sender.clone(),
         };
 
-        info!(log, "XNet Endpoint listening on {}", address);
-
         let shutdown_notify = Arc::new(Notify::new());
 
-        let shutdown = server.with_graceful_shutdown({
-            let shutdown_notify = Arc::clone(&shutdown_notify);
-            async move { shutdown_notify.notified().await }
-        });
+        let address = runtime_handle.block_on(start_server(
```

Comment:
I am surprised to see block_on here. 

Most likely you want just to enter the runtime inside the start_server function

Reply:
Yup you are right